File: src/checkers/env/encoding.py
"""Canonical eight-plane observations for the checkers environment."""

from __future__ import annotations

from collections.abc import Iterator

import numpy as np
from numpy.typing import NDArray

from checkers.rules.board import BOARD_SIZE as _BOARD_SIZE
from checkers.rules.board import coord, rotate_square
from checkers.rules.state import State

BOARD_SIZE = _BOARD_SIZE
OBSERVATION_PLANES = 8
NO_PROGRESS_LIMIT = 40
DEFAULT_MAX_PLIES = 512

Float32Array = NDArray[np.float32]


def _iter_squares(mask: int) -> Iterator[int]:
    remaining = mask
    while remaining:
        least_significant = remaining & -remaining
        yield least_significant.bit_length() - 1
        remaining ^= least_significant


def _canonical_square(state: State, square: int) -> int:
    return square if int(state.side_to_move) == 0 else rotate_square(square)


def _write_mask(observation: Float32Array, plane: int, state: State, mask: int) -> None:
    for square in _iter_squares(mask):
        row, column = coord(_canonical_square(state, square))
        observation[plane, row, column] = np.float32(1.0)


def encode_observation(
    state: State,
    *,
    max_plies: int = DEFAULT_MAX_PLIES,
) -> Float32Array:
    """Encode a state from its actor's canonical perspective.

    Args:
        state: Complete immutable state to encode.
        max_plies: Declared R6.5 terminal-ply limit used to normalize plane 7.

    Returns:
        A float32 array with shape ``(8, 8, 8)`` following the public experiment contract.

    Raises:
        TypeError: If ``state`` is not a ``State`` or ``max_plies`` is not an integer.
        ValueError: If ``max_plies`` is less than one.
    """

    if not isinstance(state, State):
        raise TypeError("state must be a State")
    if isinstance(max_plies, bool) or not isinstance(max_plies, int):
        raise TypeError("max_plies must be an integer")
    if max_plies < 1:
        raise ValueError("max_plies must be at least one")

    observation = np.zeros(
        (OBSERVATION_PLANES, BOARD_SIZE, BOARD_SIZE),
        dtype=np.float32,
    )
    actor = int(state.side_to_move)
    opponent = int(state.side_to_move.opponent)
    _write_mask(observation, 0, state, state.men[actor])
    _write_mask(observation, 1, state, state.kings[actor])
    _write_mask(observation, 2, state, state.men[opponent])
    _write_mask(observation, 3, state, state.kings[opponent])
    _write_mask(observation, 4, state, state.captured_pending)
    if state.moving_square is not None:
        _write_mask(observation, 5, state, 1 << state.moving_square)
    observation[6].fill(np.float32(state.no_progress[actor] / NO_PROGRESS_LIMIT))
    observation[7].fill(np.float32(state.ply / max_plies))
    return observation

File: web/checkers/src/engine/encoding.ts
/**
 * Actor-canonical observations and the fixed 128-slot action space.
 *
 * Ports of `checkers.env.encoding.encode_observation` and `checkers.env.masking`. The observation
 * is eight `8 × 8` planes from the side to move's perspective: own men, own kings, opponent men,
 * opponent kings, pending captures, the forced moving piece, the actor's no-progress counter, and
 * the normalized ply.
 */
import {
  BOARD_SIZE,
  DEFAULT_MAX_PLIES,
  DIRECTION_DELTAS,
  NO_PROGRESS_LIMIT,
  PLAYABLE_SQUARES,
  coord,
  hasBit,
  legalSteps,
  opponent,
  rotateSquare,
  type State,
  type Step,
} from "./rules";

export const OBSERVATION_PLANES = 8;
export const OBSERVATION_SIZE = OBSERVATION_PLANES * BOARD_SIZE * BOARD_SIZE;
export const DIRECTIONS_PER_SQUARE = 4;
export const ACTION_COUNT = PLAYABLE_SQUARES * DIRECTIONS_PER_SQUARE;

export interface LegalAction {
  readonly action: number;
  readonly step: Step;
}

function canonicalSquare(state: State, square: number): number {
  return state.side === 0 ? square : rotateSquare(square);
}

function writeMask(observation: Float32Array, plane: number, state: State, mask: number) {
  for (let square = 0; square < PLAYABLE_SQUARES; square += 1) {
    if (!hasBit(mask, square)) continue;
    const [row, column] = coord(canonicalSquare(state, square));
    observation[plane * 64 + row * BOARD_SIZE + column] = 1;
  }
}

/** Encode `state` as a `(8, 8, 8)` float32 tensor in plane-major order. */
export function encodeObservation(state: State, maxPlies = DEFAULT_MAX_PLIES): Float32Array {
  const observation = new Float32Array(OBSERVATION_SIZE);
  const actor = state.side;
  const other = opponent(actor);
  writeMask(observation, 0, state, state.men[actor]);
  writeMask(observation, 1, state, state.kings[actor]);
  writeMask(observation, 2, state, state.men[other]);
  writeMask(observation, 3, state, state.kings[other]);
  writeMask(observation, 4, state, state.capturedPending);
  if (state.movingSquare !== null) writeMask(observation, 5, state, 1 << state.movingSquare);
  observation.fill(Math.fround(state.noProgress[actor] / NO_PROGRESS_LIMIT), 6 * 64, 7 * 64);
  observation.fill(Math.fround(state.ply / maxPlies), 7 * 64, 8 * 64);
  return observation;
}

function worldDirection(step: Step): number {
  const [originRow, originColumn] = coord(step.origin);
  const [destinationRow, destinationColumn] = coord(step.destination);
  const distance = step.captured === null ? 1 : 2;
  const rowDelta = (destinationRow - originRow) / distance;
  const columnDelta = (destinationColumn - originColumn) / distance;
  return DIRECTION_DELTAS.findIndex(([dr, dc]) => dr === rowDelta && dc === columnDelta);
}

/** Encode one legal step as its canonical action ID for the side to move. */
export function stepToAction(state: State, step: Step): number {
  const direction = worldDirection(step);
  if (state.side === 1) {
    return rotateSquare(step.origin) * DIRECTIONS_PER_SQUARE + (DIRECTIONS_PER_SQUARE - 1 - direction);
  }
  return step.origin * DIRECTIONS_PER_SQUARE + direction;
}

/** Legal actions in the engine's deterministic legal-step order. */
export function legalActions(state: State): LegalAction[] {
  return legalSteps(state).map((step) => ({ action: stepToAction(state, step), step }));
}
